File: src/scoreboard/PlaceholderRegistry.ts
import { Player } from "../types/hytale";

export type PlaceholderContext = {
  player: Player;
  online: number;
  max: number;
  gameName: string;
  data: Record<string, string | number>;
};

export type PlaceholderResolver = (context: PlaceholderContext) => string | number;

export class PlaceholderRegistry {
  private readonly resolvers = new Map<string, PlaceholderResolver>();

  register(placeholder: string, resolver: PlaceholderResolver): void {
    this.resolvers.set(placeholder, resolver);
  }

  resolveLine(line: string, context: PlaceholderContext): string {
    let resolved = line;
    for (const [placeholder, resolver] of this.resolvers.entries()) {
      const value = resolver(context);
      resolved = resolved.replaceAll(`{${placeholder}}`, String(value));
    }
    return resolved;
  }
}

File: src/main/java/com/hytale/scoreboard/scoreboard/PlaceholderRegistry.java
package com.hytale.scoreboard.scoreboard;

import com.hytale.scoreboard.types.Player;
import java.util.HashMap;
import java.util.Map;

public class PlaceholderRegistry {
  public interface PlaceholderResolver {
    Object resolve(PlaceholderContext context);
  }

  public static class PlaceholderContext {
    private final Player player;
    private final int online;
    private final int max;
    private final String gameName;
    private final Map<String, Object> data;

    public PlaceholderContext(Player player, int online, int max, String gameName, Map<String, Object> data) {
      this.player = player;
      this.online = online;
      this.max = max;
      this.gameName = gameName;
      this.data = Map.copyOf(data);
    }

    public Player getPlayer() {
      return player;
    }

    public int getOnline() {
      return online;
    }

    public int getMax() {
      return max;
    }

    public String getGameName() {
      return gameName;
    }

    public Map<String, Object> getData() {
      return data;
    }
  }

  private final Map<String, PlaceholderResolver> resolvers = new HashMap<>();

  public void register(String placeholder, PlaceholderResolver resolver) {
    resolvers.put(placeholder, resolver);
  }

  public String resolveLine(String line, PlaceholderContext context) {
    String resolved = line;
    for (Map.Entry<String, PlaceholderResolver> entry : resolvers.entrySet()) {
      Object value = entry.getValue().resolve(context);
      resolved = resolved.replace("{" + entry.getKey() + "}", String.valueOf(value));
    }
    return resolved;
  }
}
